disperser/controller/encoding_manager.go

```diff
@@ -0,0 +1,200 @@
+	// NumEncodingRetries defines how many times the encoding will be retried
+	NumEncodingRetries int
+	// NumRelayAssignment defines how many relays will be assigned to a blob
+	NumRelayAssignment uint16
+	// AvailableRelays is a list of available relays
+	AvailableRelays []corev2.RelayKey
+}
+
+// EncodingManager is responsible for pulling queued blobs from the blob
+// metadata store periodically and encoding them. It receives the encoder responses
+// and creates BlobCertificates.
+type EncodingManager struct {
+	EncodingManagerConfig
+
+	// components
+	blobMetadataStore *blobstore.BlobMetadataStore
+	pool              common.WorkerPool
+	encodingClient    disperser.EncoderClientV2
+	chainReader       core.Reader
+	logger            logging.Logger
+
+	// state
+	lastUpdatedAt uint64
+}
+
+func NewEncodingManager(
+	config EncodingManagerConfig,
+	blobMetadataStore *blobstore.BlobMetadataStore,
+	pool common.WorkerPool,
+	encodingClient disperser.EncoderClientV2,
+	chainReader core.Reader,
+	logger logging.Logger,
+) (*EncodingManager, error) {
+	if int(config.NumRelayAssignment) > len(config.AvailableRelays) {
+		return nil, fmt.Errorf("NumRelayAssignment (%d) cannot be greater than NumRelays (%d)", config.NumRelayAssignment, len(config.AvailableRelays))
+	}
+	return &EncodingManager{
+		EncodingManagerConfig: config,
+		blobMetadataStore:     blobMetadataStore,
+		pool:                  pool,
+		encodingClient:        encodingClient,
+		chainReader:           chainReader,
+		logger:                logger.With("component", "EncodingManager"),
+
+		lastUpdatedAt: 0,
+	}, nil
+}
+
+func (e *EncodingManager) Start(ctx context.Context) error {
+	go func() {
+		ticker := time.NewTicker(e.PullInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				err := e.HandleBatch(ctx)
+				if err != nil {
+					if errors.Is(err, errNoBlobsToEncode) {
+						e.logger.Warn("no blobs to encode")
+					} else {
+						e.logger.Error("failed to process a batch", "err", err)
+					}
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (e *EncodingManager) HandleBatch(ctx context.Context) error {
+	// Get a batch of blobs to encode
+	blobMetadatas, err := e.blobMetadataStore.GetBlobMetadataByStatus(ctx, v2.Queued, e.lastUpdatedAt)
+	if err != nil {
+		return err
+	}
+
+	if len(blobMetadatas) == 0 {
+		return errNoBlobsToEncode
+	}
+
+	currentBlockNumber, err := e.chainReader.GetCurrentBlockNumber(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get current block number: %w", err)
+	}
+
+	for _, blob := range blobMetadatas {
+		blobKey, err := blob.BlobHeader.BlobKey()
+		if err != nil {
+			e.logger.Error("failed to get blob key", "err", err, "requestedAt", blob.RequestedAt, "paymentMetadata", blob.BlobHeader.PaymentMetadata)
+			continue
+		}
+		e.lastUpdatedAt = blob.UpdatedAt
+
+		// Encode the blobs
+		e.pool.Submit(func() {
+			for i := 0; i < e.NumEncodingRetries+1; i++ {
```

Comment:
I don't really love how this all uses one retry loop. If we are just failing to update dynamo on the last step, do we really need to reencode the blob?

Reply:
The plan was to keep encoder request idempotent (i.e. encoder checks if the blob has been encoded and returns immediately in that case). Then the extra encoding request is trivial

Reply:
Ah, ok 👍

Reply:
In order to keep the encoder request idempotent that means we do something like ListObject on s3 and make sure stuff exists there?

cc: @cody-littley we were discussing this part in the morning.

Reply:
Hmm, that could be pretty slow. This seems a bit unideal. @ian-shim , do you have thoughts?

Reply:
We don't have to list whole bucket. Since we already know the object names, we can check if those [objects exist](https://docs.aws.amazon.com/cli/latest/reference/s3api/head-object.html) right?